esp-wifi/src/preempt/mod.rs

```diff
@@ -1,49 +1,102 @@
-macro_rules! sum {
-    ($h:expr) => ($h);
-    ($h:expr, $($t:expr),*) =>
-        ($h + sum!($($t),*));
-}
+#[cfg_attr(target_arch = "riscv32", path = "preempt_riscv.rs")]
+#[cfg_attr(target_arch = "xtensa", path = "preempt_xtensa.rs")]
+pub mod arch_specific;
+
+use arch_specific::*;
+use esp_wifi_sys::include::malloc;
+
+use crate::{compat::malloc::free, hal::trapframe::TrapFrame, memory_fence::memory_fence};
+
+static mut CTX_NOW: *mut Context = core::ptr::null_mut();
+
+static mut SCHEDULED_TASK_TO_DELETE: *mut Context = core::ptr::null_mut();
 
-macro_rules! task_stack {
-    ($($task_stack_size:literal),+) => {
-        const TASK_COUNT: usize = [$($task_stack_size),+].len();
-        const TASK_STACK_SIZE: [usize; TASK_COUNT] = [$($task_stack_size),+];
-        const TOTAL_STACK_SIZE: usize = sum!($($task_stack_size),+);
-        const MAX_TASK: usize = TASK_COUNT + 1; // +1 for the user program
+pub fn allocate_main_task() -> *mut Context {
+    critical_section::with(|_| unsafe {
+        if !CTX_NOW.is_null() {
+            panic!("Tried to allocate main task multiple times");
+        }
 
-        static mut TASK_STACK: [u8; TOTAL_STACK_SIZE] = [0u8; TOTAL_STACK_SIZE];
-    };
+        let ptr = malloc(size_of::<Context>() as u32) as *mut Context;
+        core::ptr::write_bytes(ptr, 0, 1);
+        (*ptr).next = ptr;
+        CTX_NOW = ptr;
+        ptr
+    })
 }
 
-static mut TASK_TOP: usize = 1;
-static mut CTX_NOW: usize = 0;
+fn allocate_task() -> *mut Context {
+    critical_section::with(|_| unsafe {
+        if CTX_NOW.is_null() {
+            panic!("Called `allocate_task` before allocating main task");
+        }
 
-fn allocate_task() -> usize {
-    unsafe {
-        let i = TASK_TOP - 1;
-        CTX_NOW = TASK_TOP;
-        TASK_TOP += 1;
-        i
-    }
+        let ptr = malloc(size_of::<Context>() as u32) as *mut Context;
+        core::ptr::write_bytes(ptr, 0, 1);
+        (*ptr).next = (*CTX_NOW).next;
+        (*CTX_NOW).next = ptr;
+        ptr
+    })
 }
 
 fn next_task() {
     unsafe {
-        CTX_NOW = (CTX_NOW + 1) % TASK_TOP;
+        CTX_NOW = (*CTX_NOW).next;
     }
 }
 
-pub fn current_task() -> usize {
+fn remove_task(task: *mut Context) {
```

Comment:
Small bikeshed, but imo `free_task` or `delete_task` is better. It's also more consistent with the `schedule_task_deletion` method.

Reply:
A comment to describe that it will free the task memory would also be welcome.